File: ios/ReminderHomeWidgetExtension/ReminderHomeWidgetSnapshot.swift
import Foundation

struct ReminderHomeWidgetSnapshot: Decodable {
  let schemaVersion: Int
  let updatedAt: Date?
  let selectedTab: String
  let tabs: [ReminderHomeWidgetTab]

  enum CodingKeys: String, CodingKey {
    case schemaVersion
    case updatedAt
    case selectedTab
    case tabs
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    schemaVersion = try container.decodeIfPresent(Int.self, forKey: .schemaVersion) ?? 1
    let updatedAtSource = try container.decodeIfPresent(String.self, forKey: .updatedAt)
    updatedAt = ReminderHomeWidgetDateParser.parse(updatedAtSource)
    selectedTab = try container.decodeIfPresent(String.self, forKey: .selectedTab) ?? ReminderHomeWidgetShared.needsHandling
    tabs = try container.decodeIfPresent([ReminderHomeWidgetTab].self, forKey: .tabs) ?? []
  }

  func tab(_ id: String) -> ReminderHomeWidgetTab? {
    tabs.first { $0.id == id }
  }
}

struct ReminderHomeWidgetTab: Decodable, Identifiable {
  let id: String
  let label: String
  let count: Int
  let entries: [ReminderHomeWidgetEntry]
}

struct ReminderHomeWidgetEntry: Decodable, Identifiable {
  var id: String { entryId }

  let entryId: String
  let type: String
  let targetId: Int?
  let actionRecordId: Int?
  let title: String
  let statusText: String
  let displayIcon: String?
  let buttonText: String?
  let action: String?
  let canAct: Bool
}

enum ReminderHomeWidgetSnapshotState {
  case ready(ReminderHomeWidgetSnapshot)
  case unavailable(ReminderHomeWidgetSnapshotProblem)
}

enum ReminderHomeWidgetSnapshotProblem {
  case missing
  case corrupt
  case unsupportedSchema
  case stale

  var title: String {
    switch self {
    case .missing: return "尚未有 widget 資料"
    case .corrupt: return "widget 資料讀取失敗"
    case .unsupportedSchema: return "widget 資料需要更新"
    case .stale: return "widget 資料已過期"
    }
  }
}

enum ReminderHomeWidgetSnapshotReader {
  static let supportedSchemaVersion = 1
  static let staleInterval: TimeInterval = 24 * 60 * 60

  static func read(now: Date = Date()) -> ReminderHomeWidgetSnapshotState {
    guard let url = ReminderHomeWidgetShared.snapshotURL else {
      return .unavailable(.missing)
    }
    guard FileManager.default.fileExists(atPath: url.path) else {
      return .unavailable(.missing)
    }

    do {
      let data = try Data(contentsOf: url)
      let snapshot = try JSONDecoder().decode(ReminderHomeWidgetSnapshot.self, from: data)
      guard snapshot.schemaVersion == supportedSchemaVersion else {
        return .unavailable(.unsupportedSchema)
      }
      guard let updatedAt = snapshot.updatedAt, now.timeIntervalSince(updatedAt) <= staleInterval else {
        return .unavailable(.stale)
      }
      return .ready(snapshot)
    } catch {
      return .unavailable(.corrupt)
    }
  }
}

enum ReminderHomeWidgetDateParser {
  static func parse(_ source: String?) -> Date? {
    guard let source, !source.isEmpty else { return nil }
    if let date = ISO8601DateFormatter().date(from: source) {
      return date
    }
    for format in ["yyyy-MM-dd'T'HH:mm:ss.SSSSSS", "yyyy-MM-dd'T'HH:mm:ss.SSS", "yyyy-MM-dd'T'HH:mm:ss"] {
      let formatter = DateFormatter()
      formatter.locale = Locale(identifier: "en_US_POSIX")
      formatter.timeZone = TimeZone.current
      formatter.dateFormat = format
      if let date = formatter.date(from: source) {
        return date
      }
    }
    return nil
  }
}

File: android/app/src/main/kotlin/com/example/reminder_app/ReminderHomeWidgetSnapshot.kt
package com.example.reminder_app

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.time.Duration
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeParseException

data class ReminderHomeWidgetSnapshot(
    val schemaVersion: Int,
    val updatedAt: Instant?,
    val selectedTab: String,
    val tabs: List<ReminderHomeWidgetTab>,
) {
    fun tab(id: String): ReminderHomeWidgetTab? = tabs.firstOrNull { it.id == id }
}

data class ReminderHomeWidgetTab(
    val id: String,
    val label: String,
    val count: Int,
    val entries: List<ReminderHomeWidgetEntry>,
)

data class ReminderHomeWidgetEntry(
    val entryId: String,
    val type: String,
    val targetId: Int?,
    val actionRecordId: Int?,
    val title: String,
    val statusText: String,
    val displayIcon: String?,
    val buttonText: String?,
    val action: String?,
    val canAct: Boolean,
    val isRemoteBacked: Boolean,
    val syncLabel: String?,
    val syncStatus: String,
    val hasPendingMutation: Boolean,
    val pendingAction: String?,
    val actionDisabledReason: String?,
)

sealed class ReminderHomeWidgetSnapshotResult {
    data class Ready(val snapshot: ReminderHomeWidgetSnapshot) : ReminderHomeWidgetSnapshotResult()
    data class Unavailable(val reason: ReminderHomeWidgetSnapshotProblem) : ReminderHomeWidgetSnapshotResult()
}

enum class ReminderHomeWidgetSnapshotProblem {
    Missing,
    Corrupt,
    UnsupportedSchema,
    Stale,
}

object ReminderHomeWidgetSnapshotReader {
    private const val snapshotFileName = "home_widget_snapshot.json"
    private const val supportedSchemaVersion = 1
    private val staleAfter = Duration.ofHours(24)

    fun read(context: Context, now: Instant = Instant.now()): ReminderHomeWidgetSnapshotResult {
        val file = File(context.filesDir, snapshotFileName)
        if (!file.exists()) {
            return ReminderHomeWidgetSnapshotResult.Unavailable(
                ReminderHomeWidgetSnapshotProblem.Missing,
            )
        }

        val snapshot = try {
            parse(file.readText())
        } catch (_: Exception) {
            return ReminderHomeWidgetSnapshotResult.Unavailable(
                ReminderHomeWidgetSnapshotProblem.Corrupt,
            )
        }

        if (snapshot.schemaVersion != supportedSchemaVersion) {
            return ReminderHomeWidgetSnapshotResult.Unavailable(
                ReminderHomeWidgetSnapshotProblem.UnsupportedSchema,
            )
        }

        val updatedAt = snapshot.updatedAt
        if (updatedAt == null || Duration.between(updatedAt, now) > staleAfter) {
            return ReminderHomeWidgetSnapshotResult.Unavailable(
                ReminderHomeWidgetSnapshotProblem.Stale,
            )
        }

        return ReminderHomeWidgetSnapshotResult.Ready(snapshot)
    }

    fun parse(source: String): ReminderHomeWidgetSnapshot {
        val json = JSONObject(source)
        return ReminderHomeWidgetSnapshot(
            schemaVersion = json.optInt("schemaVersion", supportedSchemaVersion),
            updatedAt = parseInstant(json.optString("updatedAt", "")),
            selectedTab = json.optString("selectedTab", ReminderHomeWidgetTabs.needsHandling),
            tabs = parseTabs(json.optJSONArray("tabs") ?: JSONArray()),
        )
    }

    private fun parseTabs(tabsJson: JSONArray): List<ReminderHomeWidgetTab> {
        return buildList {
            for (index in 0 until tabsJson.length()) {
                val tabJson = tabsJson.optJSONObject(index) ?: continue
                val id = tabJson.optString("id", "")
                add(
                    ReminderHomeWidgetTab(
                        id = id,
                        label = tabJson.optString("label", ReminderHomeWidgetTabs.labelFor(id)),
                        count = tabJson.optInt("count", 0),
                        entries = parseEntries(tabJson.optJSONArray("entries") ?: JSONArray()),
                    ),
                )
            }
        }
    }

    private fun parseEntries(entriesJson: JSONArray): List<ReminderHomeWidgetEntry> {
        return buildList {
            for (index in 0 until entriesJson.length()) {
                val entryJson = entriesJson.optJSONObject(index) ?: continue
                add(
                    ReminderHomeWidgetEntry(
                        entryId = entryJson.optString("entryId", ""),
                        type = entryJson.optString("type", ""),
                        targetId = entryJson.optionalInt("targetId"),
                        actionRecordId = entryJson.optionalInt("actionRecordId"),
                        title = entryJson.optString("title", ""),
                        statusText = entryJson.optString("statusText", ""),
                        displayIcon = entryJson.optionalString("displayIcon"),
                        buttonText = entryJson.optionalString("buttonText"),
                        action = entryJson.optionalString("action"),
                        canAct = entryJson.optBoolean("canAct", false),
                        isRemoteBacked = entryJson.optBoolean("isRemoteBacked", false),
                        syncLabel = entryJson.optionalString("syncLabel"),
                        syncStatus = entryJson.optString("syncStatus", "none"),
                        hasPendingMutation = entryJson.optBoolean("hasPendingMutation", false),
                        pendingAction = entryJson.optionalString("pendingAction"),
                        actionDisabledReason = entryJson.optionalString("actionDisabledReason"),
                    ),
                )
            }
        }
    }

    private fun parseInstant(value: String): Instant? {
        if (value.isBlank()) {
            return null
        }
        return try {
            Instant.parse(value)
        } catch (_: DateTimeParseException) {
            try {
                LocalDateTime.parse(value).atZone(ZoneId.systemDefault()).toInstant()
            } catch (_: DateTimeParseException) {
                null
            }
        }
    }

    private fun JSONObject.optionalString(name: String): String? {
        if (!has(name) || isNull(name)) {
            return null
        }
        return optString(name)
    }

    private fun JSONObject.optionalInt(name: String): Int? {
        if (!has(name) || isNull(name)) {
            return null
        }
        return optInt(name)
    }
}

object ReminderHomeWidgetTabs {
    const val needsHandling = "needsHandling"
    const val attention = "attention"
    const val todayCompleted = "todayCompleted"

    val all = listOf(needsHandling, attention, todayCompleted)

    fun isKnown(value: String): Boolean = value in all

    fun labelFor(value: String): String {
        return when (value) {
            needsHandling -> "需要處理"
            attention -> "要留意"
            todayCompleted -> "今天已完成"
            else -> "需要處理"
        }
    }

    fun emptyTextFor(value: String): String {
        return when (value) {
            needsHandling -> "沒有需要處理"
            attention -> "沒有要留意"
            todayCompleted -> "今天還未完成事項"
            else -> "沒有需要處理"
        }
    }
}
